File: Numeros Pseudo-Aleatorios/XorShift 32, 64 e 128 bits/128 bits/xorshift128.scala
/*
 PRNG .......: XorShift (128 bits)
 Ano ........: 2003
 Autor ......: George Marsaglia
*/

import java.util.Date

object Main {
  class XorShift128State(var seed: Long)

  def xorshift128(state: XorShift128State): Long = {
    var random = state.seed
    random ^= random << 13
    random ^= random >>> 7
    random ^= random << 17
    state.seed = random
    random & 0x7FFFFFFFFFFFFFFFL
  }

  def main(args: Array[String]): Unit = {
    val Xn1 = new XorShift128State(1234)
    val Xn2 = new XorShift128State(new Date().getTime)

    println("GERADOR DE NUMEROS PSEUDO ALEATORIOS (GNA)")
    println("PSEUDORANDOM NUMBER GENERATOR (PRNG)")
    println("PADRAO: XorShift (128 bits)")
    println("=================================================")
    println("")
    println("   Semente: 1234          Semente: TIME")
    println("   --------------------   --------------------")

    for (i <- 1 to 10) {
      println(f"${xorshift128(Xn1)}%23d${ xorshift128(Xn2)}%23d")
    }
  }
}

File: Numeros Pseudo-Aleatorios/XorShift 32, 64 e 128 bits/128 bits/xorshift128.java
/*
 PRNG .......: XorShift (128 bits)
 Ano ........: 2003
 Autor ......: George Marsaglia
*/

import java.util.Date;

class XorShift128State {
    long seed;
}

public class Main {
    static long xorshift128(XorShift128State state) {
        long random = state.seed;
        random ^= random << 11;
        random ^= random >> 8;
        random ^= random << 19;
        state.seed = random;
        return random & 0x7FFFFFFFFFFFFFFFL; 
    }

    public static void main(String[] args) {
        XorShift128State Xn1 = new XorShift128State();
        XorShift128State Xn2 = new XorShift128State();

        Xn1.seed = 1234;
        Xn2.seed = new Date().getTime();

        System.out.println("GERADOR DE NUMEROS PSEUDO ALEATORIOS (GNA)");
        System.out.println("PSEUDORANDOM NUMBER GENERATOR (PRNG)");
        System.out.println("PADRAO: XorShift (128 bits)");
        System.out.println("=================================================");
        System.out.println("");
        System.out.println("   Semente: 1234          Semente: TIME");
        System.out.println("   --------------------   --------------------");

        for (int i = 1; i <= 10; i++) {
            System.out.printf("%23d%23d%n", xorshift128(Xn1), xorshift128(Xn2));
        }
    }
}
